src/NServiceBus.Transport.Msmq/MsmqUtilities.cs

```diff
@@ -47,28 +47,25 @@ static MsmqAddress GetIndependentAddressForQueue(MessageQueue q)
         public static Dictionary<string, string> ExtractHeaders(Message msmqMessage)
         {
             var headers = DeserializeMessageHeaders(msmqMessage);
-
+            
             //note: we can drop this line when we no longer support interop btw v3 + v4
-            if (msmqMessage.ResponseQueue != null)
+            if (msmqMessage.ResponseQueue != null && headers.TryGetValue(Headers.ReplyToAddress, out var _) == false)
             {
                 headers[Headers.ReplyToAddress] = GetIndependentAddressForQueue(msmqMessage.ResponseQueue).ToString();
             }
 
-            if (Enum.IsDefined(typeof(MessageIntentEnum), msmqMessage.AppSpecific))
+            if (Enum.IsDefined(typeof(MessageIntentEnum), msmqMessage.AppSpecific) && headers.TryGetValue(Headers.MessageIntent, out var _) == false)
```

Comment:
Since you don't care about the value I think `!headers.ContainsKey(...)` is more correct?